refactor(forms): match getError and hasError to get method signature

Toxicable opened the pull request:
closes #19734

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Feature
```

## What is the current behavior?
`getError` and `hasError` mismatch the api of `get`

Issue Number: #19734


## What is the new behavior?
apis match


## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

kara reviewed (commented):
Looks good, but can we also get a test for `hasError`, since we are changing it here?

Toxicable commented:
@kara done

kara reviewed (approved):
LGTM

kara commented:
@IgorMinar API review?

Toxicable commented on packages/forms/src/model.ts:
```diff
@@ -537,7 +537,9 @@ export abstract class AbstractControl {
    * If no path is given, it checks for the error on the present control.
    */
-  hasError(errorCode: string, path?: string[]): boolean { return !!this.getError(errorCode, path); }
+  hasError(errorCode: string, path?: Array<string|number>|string): boolean {
```
Seams valid if you're traversing through a form array.
Ill add some tests.

Toxicable commented on packages/forms/src/model.ts:
```diff
@@ -537,7 +537,9 @@ export abstract class AbstractControl {
    *
    * If no path is given, it checks for the error on the present control.
```
Wasn't sure what to change here so I added a note on how the path can be used.
Let me know if you want something else

kara replied:
@Toxicable I took a look again, but wasn't able to find the docs that you added. Is the commit up to date?

Toxicable replied:
@kara Uhhh, I guess not sorry, Ill double check this afternoon

Toxicable replied:
@kara Added some detail on what the path is meant to be and how to use it

kara replied:
Looks good. I'll let @IgorMinar take one more look though, while it's presubmitting.

Toxicable commented:
Travis failure looks like a flake to me

kara commented:
[presubmit](http://test/OCL:186625804:BASE:186625815:1519319672973:9db50399)

IgorMinar commented:
thanks

kara commented:
Some Google test failures that need to be investigated

rafaelss95 commented:
@kara sry to ping you, but do you have any news?

kara commented:
@Toxicable Would you mind rebasing this? Sorry it's been sitting so long. If you can rebase, we can take another look at the test failures and see if they're still relevant.

Toxicable commented:
@kara rebased! 
`ci/circleci: test_aio_local ` looks like a flake

kara commented:
[new presubmit](https://test.corp.google.com/ui#id=OCL:223420403:BASE:223420525:1543533441617:26cec8df)

kara commented on packages/forms/src/model.ts:
```diff
@@ -664,7 +664,9 @@ export abstract class AbstractControl {
    * @returns True when the control with the given path has the error, otherwise false.
    */
-  hasError(errorCode: string, path?: string[]): boolean { return !!this.getError(errorCode, path); }
+  hasError(errorCode: string, path?: Array<string|number>|string): boolean {
```
Can you update the type signature on `AbstractControlDirective` as well? Looks like it's inconsistent.

Toxicable replied:
oh yes, will get on that

Toxicable commented:
@kara Hey, sorry the turn around from my side has been so slow.
updated those signatures aswell and re-rebased, let me know if I missed anything else

kara reviewed (approved):
LGTM. This last change fixed the internal failures (and don't worry about timing, you are still doing better than we are :) ).

kara commented:
@IgorMinar I think it needs one more review from you since the public-API changed again.

IgorMinar reviewed (approved):
Thank you, @Toxicable!!!

IgorMinar reviewed (commented):
dang! I just noticed that this PR also needs API docs update. I asked @kara to send a follow up PR so that we don't need to block this, but in the future @Toxicable please do update the API docs.

I also noticed that you wrote a great justification of this change in the issue, but that justification is not available in the commit message, in the future it would be easier for us to review these kinds of changes if the explanation was in the commit message. thank you

IgorMinar reviewed (commented):
dang #2!

this commit should have been a feature or a fix since it changes the public api surface
otherwise the change won’t get highlighted in the changelog and nobody will know

Toxicable commented:
@IgorMinar I can change it to a fix now if you like?

kara commented:
@Toxicable If you can change it to a fix and add the justification to the commit message, that'd be awesome!

Commit:
fix(forms): match getError and hasError to get method signature

Internally getError and hasError call the AbstractControl#get method which takes  `path: Array<string | number> | string` as input, since there are different ways to traverse the AbstractControl tree.
This change matches the method signitures of all methods that use this.

Toxicable commented:
@kara done!
